shell/platform/linux/fl_task_runner.cc

```diff
@@ -0,0 +1,209 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/fl_task_runner.h"
+
+static constexpr int kMicrosecondsPerNanosecond = 1000;
+static constexpr int kMillisecondsPerMicrosecond = 1000;
+
+struct _FlTaskRunner {
+  GObject parent_instance;
+
+  FlTaskExecutor executor;
+  gpointer executor_user_data;
+
+  GMutex mutex;
+  GCond cond;
+
+  guint timeout_source_id;
+  GList /*<FlTaskRunnerTask>*/* pending_tasks;
+  gboolean blocking_main_thread;
+
+  gboolean stopped;
+};
+
+typedef struct _FlTaskRunnerTask {
+  // absolute time of task (based on g_get_monotonic_time)
+  gint64 task_time_micros;
+  FlutterTask task;
+} FlTaskRunnerTask;
+
+G_DEFINE_TYPE(FlTaskRunner, fl_task_runner, G_TYPE_OBJECT)
+
+static void fl_task_runner_dispose(GObject* object);
+
+static void fl_task_runner_class_init(FlTaskRunnerClass* klass) {
+  G_OBJECT_CLASS(klass)->dispose = fl_task_runner_dispose;
+}
+
+static void fl_task_runner_init(FlTaskRunner* self) {
+  g_mutex_init(&self->mutex);
+  g_cond_init(&self->cond);
+}
+
+void fl_task_runner_dispose(GObject* object) {
+  FlTaskRunner* self = FL_TASK_RUNNER(object);
+
+  // this should never happen because the task runner is retained while blocking
+  // main thread
+  g_assert(!self->blocking_main_thread);
+
+  g_mutex_clear(&self->mutex);
+  g_cond_clear(&self->cond);
+
+  g_list_free_full(self->pending_tasks, g_free);
+  if (self->timeout_source_id != 0) {
+    g_source_remove(self->timeout_source_id);
+  }
+
+  G_OBJECT_CLASS(fl_task_runner_parent_class)->dispose(object);
+}
+
+FlTaskRunner* fl_task_runner_new(FlTaskExecutor executor,
```

Comment:
`FlTaskExecutor` doesn't seem necessary since it's only ever used with the `FlutterEngineProcTable`. The callbacks add complexity that could be removed.

Reply:
It was there for FlTaskRunner to not have direct dependency on engine (since it itself is owned by the engine), but I can remove it.

Reply:
Actually, removing the task executor would also mean exposing `FlEngine->engine` (`FLUTTER_API_SYMBOL(FlutterEngine)`) to the task runner. That doesn't seem right.

Reply:
You can use a weak reference to the engine, which is done in a lot of the other code.

Reply:
My concern is that in order for `FlTaskRunner` to execute flutter tasks directly, I'd need to expose `FlutterEngine` (not `FlEngine`), which is currently not used anywhere else outside of `FlEngine`. 

For example on Windows, there is `TaskRunnerWin32/Uwp`, which takes `TaskExpiredCallback` responsible for executing `FlutterTask` and implemented by `FlutterWindowsEngine`.

Reply:
I removed the FlTaskExecutor [here](https://github.com/flutter/engine/pull/25884/commits/43fe73da42e70b0c669c46b9d736b5822d6244a1). Instead of exposing `FlutterEngine` there is `fl_engine_execute_task` called from `FlTaskRunner`, which holds weak reference to `FlEngine`.